Adding the details page

HoracioDuca opened the pull request:
### Added the details page of the Movies App

![image](https://user-images.githubusercontent.com/84336551/124173744-d5f26100-da81-11eb-924f-204d071e648b.png)

**Here you can found:**

• **Implemented _Navigator_**: When you click on a movie you navigate to the details page of it.

• **Movie details page created**: You can see the details of each movies fetching an API.

• **_animate_do_ package implemented:** When the app is opened the movies are shown by an animation and in the details page as well.

• **A list of stars was created by logic implementation**: the voteAverage given by the API is divided by 2 and turned from _double_ to _int_. Then the result is painted with 'yellow stars' and unreachable scores are simply white.

• **README** file was modified.


**DEMO:**

![movies2](https://user-images.githubusercontent.com/84336551/124174330-9bd58f00-da82-11eb-9544-ef6ad6797076.gif)

HoracioDuca committed:
Adding the details page

FranciscoGaraventa commented on lib/src/models/result_movie.dart:
```diff
@@ -54,20 +54,32 @@ class ResultMovie {
+      voteAverage: parsedJson['vote_average'] != null
+          ? parsedJson['vote_average']
+          : 'Not voted yet',
+      title: parsedJson['title'] != null ? parsedJson['title'] : 'No title',
```
'No title' can be added to String file

HoracioDuca replied:
Done

FranciscoGaraventa commented on lib/src/models/result_movie.dart:
```diff
@@ -54,20 +54,32 @@ class ResultMovie {
+          : 'N/A',
+      originalTitle: parsedJson['original_title'] != null
+          ? parsedJson['title']
+          : 'No title',
```
'No title' can be added to String file

HoracioDuca replied:
Done

FranciscoGaraventa commented on lib/src/models/result_movie.dart:
```diff
@@ -54,20 +54,32 @@ class ResultMovie {
-      originalTitle: parsedJson['original_title'],
+      originalLanguage: parsedJson['original_language'] != null
+          ? parsedJson['original_language']
+          : 'N/A',
```
This can be added to String too

HoracioDuca replied:
Done

FranciscoGaraventa commented on lib/src/models/result_movie.dart:
```diff
@@ -54,20 +54,32 @@ class ResultMovie {
+          : 'Not overview has been provided',
+      releaseDate: parsedJson['release_date'] != null
+          ? parsedJson['release_date']
+          : 'Unknown date',
```
This can be added to String file too

HoracioDuca replied:
done

FranciscoGaraventa commented on lib/src/ui/movie_detail.dart:
```diff
@@ -0,0 +1,156 @@
+import 'package:animate_do/animate_do.dart';
+import 'package:flutter/material.dart';
+import 'package:movies_bloc/src/utils/movie_strings.dart';
```
You missed a relative import here

HoracioDuca replied:
Done

FranciscoGaraventa commented on lib/src/models/result_movie.dart:
```diff
@@ -54,20 +54,32 @@ class ResultMovie {
+          : 'Not video was found',
+      voteAverage: parsedJson['vote_average'] != null
+          ? parsedJson['vote_average']
+          : 'Not voted yet',
```
This can be added to String file too

HoracioDuca replied:
Done

FranciscoGaraventa commented on lib/src/ui/movie_detail.dart:
```diff
@@ -0,0 +1,156 @@
+                  ),
+                ),
+              ),
+              Padding(
+                padding: const EdgeInsets.only(
+                    left: MovieDimensions.edgeInsetsLeftPadding,
+                    top: MovieDimensions.edgeInsetsTopPadding),
+                child: Text(
+                  result.overview,
+                  style: TextStyle(
+                    color: MovieColors.customWhite
+                        .withOpacity(MovieDimensions.textStyleOpacity),
+                    fontSize: MovieDimensions.subtitleFontSize,
+                    fontWeight: FontWeight.bold,
+                  ),
+                ),
+              ),
```
This can be added into a new Widget for overview

FranciscoGaraventa commented on lib/src/ui/movie_detail.dart:
```diff
@@ -0,0 +1,156 @@
+                padding: const EdgeInsets.only(
+                  left: MovieDimensions.edgeInsetsLeftPadding,
+                ),
+                child: Text(
+                  MovieStrings.overview,
+                  style: TextStyle(
+                    color: MovieColors.customWhite.withOpacity(
+                      MovieDimensions.overviewTextOpacity,
+                    ),
+                    fontSize: MovieDimensions.overviewTextFontSize,
+                    fontWeight: FontWeight.bold,
+                  ),
+                ),
+              ),
+              Padding(
+                padding: const EdgeInsets.only(
+                    left: MovieDimensions.edgeInsetsLeftPadding,
+                    top: MovieDimensions.edgeInsetsTopPadding),
+                child: Text(
+                  result.overview,
+                  style: TextStyle(
+                    color: MovieColors.customWhite
+                        .withOpacity(MovieDimensions.textStyleOpacity),
+                    fontSize: MovieDimensions.subtitleFontSize,
+                    fontWeight: FontWeight.bold,
+                  ),
+                ),
+              ),
+            ],
+          ),
+        ),
+      ),
```
This can be added into a new Widget for overview

HoracioDuca replied:
Done, thank you

FranciscoGaraventa commented on lib/src/ui/movie_detail.dart:
```diff
@@ -0,0 +1,156 @@
+import '../utils/movie_dimensions.dart';
+import '../models/result_movie.dart';
+
+class MovieDetail extends StatelessWidget {
```
Here you can abstract a few Widgets

HoracioDuca replied:
Done

HoracioDuca committed:
String texts moved to strings file, corrected relative imports and made the widgets abstractions

voteAverage is now var type, movieSearchBar was changed as expected and
whole bloc passing through changed

Removed toString from posterPath, movieReleaseDate is now late String
type, corrected missed relative imports and refactored movies_grid_view

buildList abstracted to a new file called movie_grid_builder

Added a trailing comma that were missing.

the new indicator was removed and the name of the method was modified to
be more discreptive.

FlexibleSpace now receive a child widget and updated null operators in
result_movie file.

FranciscoGaraventa commented on lib/src/models/result_movie.dart:
```diff
@@ -1,10 +1,10 @@
   late int id;
   bool? video;
-  var voteAverage;
+  dynamic voteAverage;
```
Why voteAverage is dynamic?

HoracioDuca replied:
I think it is a dynamic type, because when I use it to show voteAverage I need it to be String type and when I need to do the logic to show the list of stars I need it to be Integer type so, what do you think that it would be better in this case? Maybe left it just in Integer? Or dynamic as I thought or var as it was?

eis6aer replied:
Whats the data type thats the server is sending in the response ?

FranciscoGaraventa replied:
In that case, the type on json said that is integer, if you need to change it to String, you can parse it.

eis6aer replied:
exactly !

HoracioDuca replied:
The type of data that server returns is 'dynamic', that's why I also changed into it

HoracioDuca replied:
Fixed, I finally lefted var type.

FranciscoGaraventa commented on lib/src/widgets/movie_appbar_search_bar.dart:
```diff
@@ -0,0 +1,40 @@
+import '../blocs/i_movies_bloc.dart';
+import '../utils/movie_strings.dart';
+
+class MovieAppBarSearchBar extends StatefulWidget {
```
What happen if you want to use MovieAppBarSearchBar elsewhere not necessarily on appBar? I think you can change this name, for MovieSearchBar, because the idea of abstract a widget is that you can use it in more than one place not only on AppBar.

HoracioDuca replied:
You are totally right. Fixed.

FranciscoGaraventa commented on lib/src/widgets/movie_body_homepage.dart:
```diff
@@ -0,0 +1,102 @@
+import '../utils/movie_dimensions.dart';
+
+class MovieBodyHomePage extends StatefulWidget {
+  final IMoviesBloc moviesBloc;
```
Instead of passing all the bloc you can use only the data inside, because you already call the method that add movies into stream on HomePage

FranciscoGaraventa replied:
Another approach that you can do is change widget name, think that here you are only creating a GridView to show movies

HoracioDuca replied:
That's a nice tip. Fixed.

FranciscoGaraventa commented on lib/src/ui/movie_detail.dart:
```diff
@@ -0,0 +1,69 @@
+            crossAxisAlignment: CrossAxisAlignment.start,
+            children: [
+              MovieDetailPoster(
+                moviePosterPath: result.posterPath.toString(),
```
Why you parse this to String, posterPath is already a String

HoracioDuca replied:
Corrected.

FranciscoGaraventa commented on lib/src/widgets/movie_detail_subtitle.dart:
```diff
@@ -0,0 +1,30 @@
+import '../utils/movie_dimensions.dart';
+
+class MovieDetailSubtitle extends StatelessWidget {
+  final String? movieReleaseDate;
```
Why this String can be null but in constructor is required?

HoracioDuca replied:
Fixed. Changed from null-able to 'late String' type

FranciscoGaraventa commented on lib/src/widgets/movies_grid_view.dart:
```diff
@@ -0,0 +1,101 @@
+import 'package:animate_do/animate_do.dart';
+import 'package:flutter/material.dart';
+import 'package:movies_bloc/src/ui/movie_detail.dart';
```
You missed a relative import here

HoracioDuca replied:
Done

FranciscoGaraventa commented on lib/src/widgets/movies_grid_view.dart:
```diff
@@ -0,0 +1,101 @@
+            );
+          },
+        )
+      : Container();
```
You can add some content to this Container, instead of an empty Container you can added an image or icon that show that something goes wrong when you doesn't have any data on the stream.

HoracioDuca replied:
movies_grid_view was refactored so Container is no longer available. Now you just consult one time if the stream hasData and create the buildList, other ways it just show the CircularProgressIndicator.

MarcosCGlobant commented on lib/src/widgets/movies_grid_view.dart:
```diff
@@ -0,0 +1,99 @@
+  }
+}
+
+Widget buildList(AsyncSnapshot snapshot) {
```
do you think this can be moved to another widget?

HoracioDuca replied:
buildList is now in 'movies_grid_builder' file, thank you.

MarcosCGlobant commented on lib/src/widgets/movie_detail_overview.dart:
```diff
@@ -0,0 +1,44 @@
+          movieOverview,
+          style: TextStyle(
+            color: MovieColors.customWhite
+                .withOpacity(MovieDimensions.textStyleOpacity),
```
missed a trailing comma here

HoracioDuca replied:
Fixed, thank you

MarcosCGlobant commented on lib/src/widgets/movie_grid_builder.dart:
```diff
@@ -0,0 +1,62 @@
+import '../ui/movie_detail.dart';
+import '../utils/movie_dimensions.dart';
+
+GridView buildList(AsyncSnapshot snapshot) {
```
you can make the name of the method a bit more descriptive, also if you are returning a GridView, you should make a reference to that

HoracioDuca replied:
done

MarcosCGlobant commented on lib/src/widgets/movie_grid_builder.dart:
```diff
@@ -0,0 +1,62 @@
+GridView buildList(AsyncSnapshot snapshot) {
+  return GridView.builder(
+    itemCount: snapshot.data.results.length,
+    gridDelegate: new SliverGridDelegateWithFixedCrossAxisCount(
```
the "new" indicator it's not necessary, please remove it

HoracioDuca replied:
done

fbarh commented on lib/src/models/result_movie.dart:
```diff
@@ -54,20 +54,32 @@ class ResultMovie {
-      video: parsedJson['video'],
-      voteAverage: parsedJson['vote_average'],
-      title: parsedJson['title'],
+      video: parsedJson['video'] != null
```
Take a look at this null operator: https://news.dartlang.org/2015/08/dart-112-released-with-null-aware.html, this can make you save coding time a it will make your code easy to read

HoracioDuca replied:
Done, thank you very much for the tip. Also about the comment above I have to say that I've tried to implement the json_serializable package but I couldn't make it work, but I'll take that advise for the future for sure, thank you for that too.

fbarh reviewed (commented):
When you have 1 or two models, it is easy to work with factory methods inside the model classes. but when the data model has more than 10 classes and they are related, this approach is hard to support. Take a look at this way/lib to work with json objects a little bit easily (it is a dev library).
https://pub.dev/packages/json_serializable

luisjavierglb commented on lib/src/ui/home_page.dart:
```diff
@@ -95,77 +78,12 @@ class _HomePageState extends State<HomePage> {
-            ),
-          ),
-        ),
+        flexibleSpace: MovieAppBarFlexibleSpaceStyle(),
```
You can make this FlexibleSpace to receive a child widget in case that you want to show something on that space, make sense?

HoracioDuca replied:
It makes sense to me, in the future I could add some 'data' in that space. Thank you.

luisjavierglb commented on lib/src/models/movie.dart:
```diff
@@ -18,16 +18,8 @@ class Movie {
-        )
-        .toList();
+    List<ResultMovie> movieTemp =
+        allMovies.map((eachMovie) => ResultMovie.fromJson(eachMovie)).toList();
```
be careful of this allMovies, it could be null

HoracioDuca replied:
Fixed, now allMovies has a checkpoint that handle if the value were null

luisjavierglb reviewed (approved):
LGTM

MarcosCGlobant reviewed (approved):
LGTM
